### ISSUE TITLE
FR variational distribution over timed trees

### ISSUE THREAD

**Issue by fritzo** [broadinstitute/pyro-cov#1]
Here is a sketch

Model:

<details>

```py
def model(leaf_times, leaf_states, num_features):
   assert len(leaf_times) == len(leaf_states)
   
   # Timed tree concerning reproductive behavior only.
   coal_params = pyro.sample("coal_params", CoalParamPrior())  # global
   # Note this is where our coalescent model assumes geographically
   # homogeneous reproductive rate, which is not realistic.
   # See appendix of (Vaughan et al. 2014) for discussion of this assumption.
   parents_and_times = pyro.sample("parents_and_times",
                                   Coalescent(coal_params, leaf_times))
   
   # This is compatible with subsampling features, but not leaves.
   subs_params = pyro.sample("subs_params", GTR_gamma_prior)  # global
   with pyro.plate("features", num_features, leaf_states.size(-1)):
       # This is similar to the phylogeographic likelihood in the pyro-cov repo.
       # This is simpler (because it is time-homogeneous)
       # but more complex in that it is batched.
       # This computes mutation likelihood via dynamic programming.
       pyro.sample("leaf_times", PhylogeneticTree(parents, times, subs_params),
                   obs=leaf_states)
```

</details>

Guide:

<details>

```py
def guide(leaf_times, leaf_states, num_features):
    assert len(leaf_times) == len(leaf_states)
   
    # Sample all continuous latents in a giant correlated auxiliary.
    aux = pyro.sample("aux", LowRankMultivariateNormal(TODO))
    # Split it up (TODO switch to EasyGuide).
    pyro.sample("coal_params", Delta(aux["TODO"]))  # global
    pyro.sample("subs_params", Delta(aux["TODO"]))  # global
    # These are the times of each bit in the embedding vector.
    bit_times = pyro.sample("bit_times", Delta(aux["TODO"]),
                            infer={"is_auxiliary": True})

    # Learn parameters of the discrete distributions,
    # possibly conditioned on continuous latents.
    if amortized:
        # Amortized guide, compatible with subsampling leaves but not features.
        logits = my_nn(leaf_states, leaf_times)  # batched over leaves
    else:
        # Fully local guide, compatible with subsampling features but not leaves.
        with pyro.plate("leaves", len(leaf_times)):
            logits = pyro.param("logits",
                lambda: torch.randn(leaf_times.shape),
                event_dim=0)
    assert len(logits) == len(leaf_times)

    pyro.sample("parents_and_times",
        VariationalTree(bit_times, logits))
```

</details>

Variational distribution with straight-through gradients:

<details>

```py
class VariationalTree(TorchDistribution):
    """
    Samples tree topology and times
    """
    has_rsample = True  # only wrt times, not parents

    def __init__(self, bit_times, logits):
        self.stuff = stuff
        
    def rsample(self, sample_shape=torch.Size()):
        if sample_shape:
            raise NotImplementedError
        bits = Bernoulli(logits=logits).sample()
        num_leaves, num_bits = bits.shape
        # TODO figure out a cheaper algorithm here.
        # TODO then ensure gradients work.
        partitions = [set(range(num_leaves))]
        for b, t in sorted(bits, key=times):
            for partition in partitions:
                sub = defaultdict(set)
                for node in partition:
                    sub[bits[node]].add(node)
                ...blah blah...
        parents = TODO
        times = TODO
        return parents, times
```

</details>

cc @eb8680
